Fixes #23714 - Bash completion for foreman-maintain

mbacovsky opened the pull request:
See README for details on how to install and use

iNecas commented on README.md:
```diff
@@ -385,6 +385,39 @@ foreman-maintain console
+
+Possible options for the `:completion` attribute are:
+ - `{}` option has no value, default for flags
+ - `{ :value => {} }` option has value of unknown type, no suggestions for the value, default
```
The empty hash in `{ :value => {} }` looks a bit odd. What about somehing like: `{ :type => :value }`, for file it could be `{ :type => :file, :filter => '\.txt$'}`

iNecas replied:
Or, for simpler cases, maybe `{ :completion => :directory }`

iNecas replied:
One suggestion - doesn't have to be solved in this PR, but would be good to think about this now: It would be great if `foreman-maintain upgrade check --target-version ` could produce a list of versions that the system can be upgraded to: since this list could be dynamic, we would need to generate it on actual setup (rather than in packaging time). I can imagine the setting by something like `{ :type => :value, :generator => lambda { Upgrades.available_versions } }` - warping to lambda so that it would not be evaluated every time the file is sourced.

mbacovsky replied:
@inecas I'll re-check on notation, what you suggest  feels more natural.
Regarding the target version value, I have some plans for this already. The cache is never build during packaging but rather during rpm %post. Then we can have 
```ruby
{ :type => :enum, :values => Upgrades.available_versions } 
```
We should probably call the cache rebuild as part of upgrade scenario.

The lambda probably won't work as we don't load whole ForemanMaintain (for performance reasons) just the completion parser, so no FM objects are available during the parsing.

Part of more ambitious plans is to add `:function` type. In the project you would inherit from the `Completion` class and implement there methods returning the more dynamic data on the fly e.g. `organizations', `products`, `capsules`, etc.. In the option you could do
```ruby
{ :type => :function, :name => :capsules }
```

iNecas replied:
> The lambda probably won't work as we don't load whole ForemanMaintain

The idea is that the lambda would be called only on cache-creation phase, to store the values in the cache, not in the completion parser.

iNecas commented on lib/foreman_maintain/utils/bash.rb:
```diff
@@ -0,0 +1,154 @@
+          end
+        end
+
+        def self.load_description(path)
```
Can be simplified to:

```ruby
def self.load_description(path)
  YAML.load(File.open(path))
rescue Errno::ENOENT
  {}
end
```

iNecas commented on lib/foreman_maintain/utils/bash.rb:
```diff
@@ -0,0 +1,154 @@
+
+        def next_param(dict)
+          if dict.key?(:params)
+            if dict[:params].first.key?(:directory)
```
The type would be either `:param` or `:option`

iNecas commented on README.md:
```diff
@@ -385,6 +385,39 @@ foreman-maintain console
+
+Bash completion is automatically installed by RPM. To use it for development setup 
+`cp ./config/foreman-maintain.completion /etc/bash_completion.d/foreman-maintain` 
+and load it to the current shell `source /etc/bash_completion.d/foreman-maintain`.
```
When following this instructions in development, I was getting:

```
foreman-maintain-complete: command not found
```
as it assumes it's on the path. We should probably mention that the the PATH needs to be set accordingly

iNecas replied:
We could also assume that in development, the executable will bi in `$PWD/bin`

iNecas commented:
Tests are currently filing.

iNecas commented:
Tested and works flawlessly. Nice work @mbacovsky. Some suggestions inline. I'm wondering if we should even bother with building the cache in build phase of packaging… the first regeneration was not that terrible. And with things like dynamically generated completion lists, we could not even do it beforehand. What we could/should do is running that in post-inst/post-upgrade phase. Maybe that was what you already meant, but just in case…

mbacovsky commented:
@sean797 that's exactly my plan. I will move it to separate gem if this concept proves itself. It could be re-used for other Clamp based apps we have in the stack i.e. hammer and installer as you suggest.

mbacovsky commented:
@iNecas for the cache build I meant to do that on RPM @post. As you said it makes no sense to do it during package build. What didn't occurred to me is the case when f-m is installed sooner then installer finishes. It may be better to omit the pre-build completely. 

We should trigger the cache build after the installer is run to update after possible changes. I'll add something to expire the cache on installer answers change.

mbacovsky commented:
@iNecas, updated.
- notation updated as suggested
- refactoring of the code
- fixed completion of files and directories to feel more bash-like
- cache expiration, cache expires when scenario answer file changed
- updated docs

I see now why the tests are failing in Jenkins, will update shortly.

iNecas commented on lib/foreman_maintain/utils/bash.rb:
```diff
@@ -0,0 +1,138 @@
+          dict.keys.select { |k| k.is_a?(String) }.map { |k| k + ' ' }
+        end
+
+        def complete_value(value, partial, is_param)
```
the `value` name is misleading: what about something like `completion_options` instead?

mbacovsky replied:
how about `value_description`?

alda519 commented on lib/foreman_maintain/cli/backup_command.rb:
```diff
@@ -44,7 +45,8 @@ def common_backup_options
           option ['-i', '--incremental'], 'PREVIOUS_BACKUP_DIR',
-                 'Backup changes since previous backup' do |dir|
+                 'Backup changes since previous backup',
+                 :completion => { :type => :file } do |dir|
```
directory type?

alda519 commented on bin/foreman-maintain-complete:
```diff
@@ -0,0 +1,46 @@
+              else
+                File.join(project_root, 'config/foreman_maintain.yml')
+              end
+completion_cache_file = cache_file(config_file) || '~/.cache/foreman_maintain_completion.yaml'
```
there are places with `yaml` and other places with `yml` file extenstion, I would make it consistent

iNecas commented:
I've hit one issue during testing:

```
foreman-maintain backup online  /var/lib/ <TAB>/root/foreman_maintain/lib/foreman_maintain/utils/bash.rb:52:in `complete_value': undefined method `[]' for nil:NilClass (NoMethodError)
	from /root/foreman_maintain/lib/foreman_maintain/utils/bash.rb:77:in `next_param'
	from /root/foreman_maintain/lib/foreman_maintain/utils/bash.rb:44:in `next_word'
	from /root/foreman_maintain/lib/foreman_maintain/utils/bash.rb:20:in `complete'
	from /root/foreman_maintain/bin/foreman-maintain-complete:44:in `<main>'

```

alda519 commented on bin/foreman-maintain-complete:
```diff
@@ -0,0 +1,46 @@
+require 'English'
+project_root = File.expand_path('../..', __FILE__)
+$LOAD_PATH.unshift(File.join(project_root, 'lib'))
+CONFIG_FILE = '/etc/foreman-maintain/foreman_maintain.yaml'.freeze
```
should be `/etc/foreman-maintain/foreman_maintain.yml`?

mbacovsky commented:
@iNecas, @alda519 updated:
- error on completion fixed
- using 'yml' instead of 'yaml'
- fixed type in --incremental option
- some more refactoring and code reduction

mbacovsky commented:
@iNecas it will need changes in packaging, probably just installing the completion file. Should I prepare the PR?

iNecas commented:
Ideally yes, if you can't make it before the time off, it will probably wait for you later, or we will try to find someone else for the packaging part.

mbacovsky commented:
@iNecas, I've updated gemspec. Incomplete attempt to update packaging can be found here [1] if needed. [1] https://github.com/mbacovsky/foreman-packaging/commit/3822d09ccc7be6dc8efb39d006bdd55918bfbc42

mbacovsky commented:
@iNecas, I've added fixes for completion issues found during testing packaged version on a fresh box.

I've also created a test build and it seems it installs fine, bash-completion works out of the box (almost, you have to load the completion defs by e.g. starting new shell). 

Not sure how to proceed with the packaging PR. We can either copy my changes when releasing 0.2.7-1 or add them in 0.2.7-2 as a separate PR. What do you think? My packaging changes are in https://github.com/mbacovsky/foreman-packaging/commit/3173c096ee202bc73f9eea26205c23e3745069d1. I was not able to do scratch build in Koji so did local mockbuild, you can check it here https://drive.google.com/open?id=1w51YNsZIdvYCy9jYPdHfYhof5KxJhOci.

iNecas commented:
Left some nitpicks. Overall reads better now. I've also tested and worked well for me: the only obstacle was the fact that you've built it for tfm software collection so that there were some packaging conflicts and I needed to `rpm -e --nodeps rubygem-foreman_maintain` first before I could install it. After that, everything worked as expected. @alda519 / @sean797 want to give it a try before merging?

Regarding packaging, I'm ok with doing the changes as part of 0.2.7-1 release.

iNecas commented on bin/foreman-maintain-complete:
```diff
@@ -0,0 +1,46 @@
+# rubocop:disable Lint/RescueWithoutErrorClass, Lint/HandleExceptions
+def cache_file(config_file)
+  config = YAML.load(File.open(config_file))
+  config[:completion_cache]
```
Missed rename?

mbacovsky commented:
@sean797 we decided not to build the completions during a build. The cache depends on configuration of the system it is used with so the build would produce outdated cache. The cache is currently built during a first attempt to complete something and is expired by installer run. The build on my testing system takes less then a second. This can fixed in the future if too obtrusive. Sounds reasonable?

iNecas commented:
@mbacovsky I would like to merge it as soon as you have time to finish the packaging part: when is the right time?

mbacovsky commented:
@iNecas I consider the packaging part finished already. It is in my branch mentioned in the previous comment. I'm not sure what I'm supposed to provide now. My expectation was there will be new f-m release after my PR is merged and during the packaging  we will add the changes from my branch. Or is the plan different?

iNecas commented:
Ok, I read it as you'll open a PR to packaging right after I merge this. To here you go :)

iNecas commented:
Thanks @mbacovsky and all involved in this beautiful change.
